Border panel: Collapse color controls

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
In #37067 the Color panel was updated to collapse the color elements for a cleaner looking/more compact set of controls. This PR updates the Border panel to do the same for the border color.

To accomplish this, it refactors out the `ColorGradientSettingsDropdown` into its own component so that it can be used by the Border panel.

| Before | After |
| --- | --- |
| <img width="266" alt="Screen Shot 2021-12-15 at 1 13 15 PM" src="https://user-images.githubusercontent.com/63313398/146265792-e92c02ae-9376-4036-a7ac-5cee3b3c3e66.png"> | <img width="273" alt="Screen Shot 2021-12-16 at 11 10 30 AM" src="https://user-images.githubusercontent.com/63313398/146437272-9da830b9-5510-4eb5-9d32-c7f81aa137c0.png"> |


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

I used the Group block to test, since it supports Color and Border with all its properties.

* Insert a Group block and verify the Border panel matches screenshots.
* Click on the Border color control and verify the popover opens. Test making selections.
* Verify that border color can be reset via the Tools panel menu, and via the Reset all button in the menu.
* Verify that the Background, Text, and Link color options in the Colors panel continue to work.

## Screenshots <!-- if applicable -->

With the popover open:

<img width="604" alt="Screen Shot 2021-12-15 at 1 19 58 PM" src="https://user-images.githubusercontent.com/63313398/146266682-58ef4901-628e-48d8-bde3-fdad5a5dc64a.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Refactor out Dropdown component to be used by border panel

stacimc committed:
Update label text

stacimc commented:
The border controls are being reimagined entirely in #35602. In the meantime, this PR greatly reduces the size of the panel by collapsing the colors.

ramonjd commented:
I tested in the post and site editors.

✅  Click on the Border color control and verify the popover opens. Test making selections.
✅  Verify that border color can be reset via the Tools panel menu, and via the Reset all button in the menu.

![Kapture 2021-12-16 at 09 09 10](https://user-images.githubusercontent.com/6458278/146272970-078cc7f1-c43b-4a8a-9efc-59d869609927.gif)

✅  Verify that the Background, Text, and Link color options in the Colors panel continue to work.

![Kapture 2021-12-16 at 09 07 21](https://user-images.githubusercontent.com/6458278/146272987-e1dc35da-32f3-42d9-859a-9cec9abef134.gif)


This is not an issue with this PR, but I noticed that the floating panel isn't quite optimized for narrow screens:
<img width="200" alt="Screen Shot 2021-12-16 at 9 06 29 am" src="https://user-images.githubusercontent.com/6458278/146273052-de672b0e-789c-41b8-815f-ee25f47e5039.png">

ramonjd reviewed (approved):
This works well for me, and is a useful abstraction IMO.

👍

jasmussen commented:
Thanks for working on this. I'm curious, the ItemGroup for the Paragraph has 2px rounded corners for the first and last items:

<img width="270" alt="Screenshot 2021-12-16 at 09 40 26" src="https://user-images.githubusercontent.com/1204802/146337299-7bf140f5-82f4-4923-9a45-0415901d69b1.png">

This one has square corners:
<img width="272" alt="Screenshot 2021-12-16 at 09 41 37" src="https://user-images.githubusercontent.com/1204802/146337460-5ee7055c-e0b1-4778-80ff-eb520d0e02bf.png">

_Note that all of these itemgroups are too tall, 43px in global styles, 46px in the post editor. This is a separate issue unrelated to this PR._

jasmussen commented:
CC: @ciampo if you have any inputs on the rounded corners.

stacimc committed:
Add border-radius

stacimc committed:
Refactor out the entire ItemGroup instead of just the Dropdown

Exports the entire ItemGroup so that other panels that want to
use the component without the wrapping PanelBody can still support
multiple color items, and do not need to manually apply border
styles. It continues to work when passed only a single item.

stacimc commented:
> I'm curious, the ItemGroup for the Paragraph has 2px rounded corners for the first and last items:
>
> This one has square corners:

@jasmussen Well spotted, thank you! The border in the Color panel is applied via the ItemGroup as you mentioned; I pulled out the single item Dropdown for use in the Border panel.

I've updated this to refactor out the entire `ItemGroup` instead, so that border styles don't have to be manually applied for components that want to use the dropdown outside of the ColorPanel. This will also allow panels to support multiple color Items if the need arises.

<img width="273" alt="Screen Shot 2021-12-16 at 11 10 30 AM" src="https://user-images.githubusercontent.com/63313398/146437215-71989b3a-7108-4072-aaff-27fd2a2f58dc.png">

andrewserong reviewed (approved):
Great work @stacimc, nice refactor! Just confirming this is testing well for me too:

✅ Panel matches the screenshots in the PR description
✅ Popover panel and selecting from custom or theme/default colors works well
✅ Border color is correctly reset via the ToolsPanel menu buttons
✅ Verified that the Background, Text, and Link color options in the Colors panel continue to work, and in the process made some fairly hilarious designs 😅

![image](https://user-images.githubusercontent.com/14988353/146466368-8bc8f402-85c9-483c-a8f9-d2776dd2605f.png)

The only issue I ran into was the same as Ramon mentions about the popover position on mobile. That issue exists on `trunk` anyway, so this looks good to go to me, and we can always do follow-ups.

stacimc commented:
Rebased to fix merge conflicts. I'm seeing scrollbars on the popover now when the screen height is small, but confirmed this is [acknowledged](https://github.com/WordPress/gutenberg/pull/37430#issuecomment-996993003) behavior and happens on trunk.
<img width="302" alt="Screen Shot 2021-12-17 at 3 55 08 PM" src="https://user-images.githubusercontent.com/63313398/146620650-f1bd9c5f-65b4-4682-8f21-d77a60993120.png">

ciampo commented on packages/block-editor/src/components/colors-gradients/dropdown.js:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	return (
+		<ItemGroup
+			isBordered
+			isSeparated
+			className="block-editor-panel-color-gradient-settings__item-group"
+		>
+			{ settings.map( ( setting, index ) => (
+				<Dropdown
+					key={ index }
+					position={ dropdownPosition }
+					className="block-editor-panel-color-gradient-settings__dropdown"
+					contentClassName="block-editor-panel-color-gradient-settings__dropdown-content"
+					renderToggle={ ( { isOpen, onToggle } ) => {
+						return (
+							<Item
```
I don't remember if this exact situation already came up in another PR (cc @jasmussen ), but have you tried to have the `ItemGroup` inside `Dropdown` instead? ie.

```jsx
<Dropdown>
  <ItemGroup>
    <Item>
```

jasmussen replied:
We did try that, yes, and there were ome problems with it. I can't recall them exactly, they are noted somewhere on one of Jorges PRs, and it's something to follow up on, but was acceptable for now. Thanks for looking, and thanks Staci!

ciampo reviewed (commented):
> CC: @ciampo if you have any inputs on the rounded corners.

Sorry for the delayed reply! I think that @stacimc already fixed the issue correctly — in fact, `Item` and `ItemGroup` are intended to be used together :)

Mamaduka commented:
Cherry-picked for 5.9.1.
